cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/handler.py

```diff
@@ -0,0 +1,833 @@
+                assetRootExtId,
+                docTypeMetaCol,
+                pAndIdDocType,
+                runAll,
+                debug,
+                docLimit,
+            )
+            entities = get_files_entities(files_all)
+
+            if len(entities) > 0:
+                annotationList = get_existing_annotations(cognite_client, entities)
+
+            numAnnotated = 0
+            numErrors = 0
+            if len(files_process) > 0:
+                entities = get_assets(cognite_client, assetRootExtId, entities)
+                numAnnotated, numErrors = process_files(
+                    cognite_client,
+                    matchTreshold,
+                    entities,
+                    files_process,
+                    annotationList,
+                    debug,
+                )
+
+            msg = f"Annotated P&ID files for asset: {assetRootExtId} number of files annotated: {numAnnotated} - file not annotaded due to errors: {numErrors}"
+            logger.info(msg)
+            cognite_client.extraction_pipelines.runs.create(
+                ExtractionPipelineRun(
+                    extpipe_external_id=extractionPipelineExtId,
+                    status="success",
+                    message=msg,
+                )
+            )
+
+        except (CogniteAPIError, Exception) as e:
+            msg = f"Annotated P&ID files failed on root asset: {assetRootExtId} failed - Message: {e!s}"
+            logger.exception(msg)
+            if len(msg) > 1000:
+                msg = msg[0:995] + "..."
+            cognite_client.extraction_pipelines.runs.create(
+                ExtractionPipelineRun(
+                    extpipe_external_id=extractionPipelineExtId,
+                    status="failure",
+                    message=msg,
+                )
+            )
+            pass
+
+
+def get_files(
+    cognite_client: CogniteClient,
+    docDataSetExtId: str,
+    assetRootExtId: str,
+    docTypeMetaCol: str,
+    pAndIdDocType: str,
+    runAll: bool,
+    debug: bool,
+    docLimit: int = -1,
+) -> tuple[dict[str, FileMetadata], dict[str, FileMetadata]]:
+    """
+    Read files based on doc_type and mime_type to find P&ID files
+
+    :param cognite_client:
+    :param pAndIdDocType:
+    :param runAll:
+    :param debug:
+    :param docLimit:
+
+    :returns: dict of files
+    """
+
+    pAndIdFiles_all: dict[str, FileMetadata] = {}  # Define a type for Dict
+    pAndIdFiles_process: dict[str, FileMetadata] = {}  # Define a type for Dict
+    numDoc = 0
+    metaFileUpdate: list[FileMetadataUpdate] = []
+
+    logger.info(
+        f"Get files to annotate data set: {docDataSetExtId}, asset root: {assetRootExtId} doc_type: {pAndIdDocType} and mime_type: {orgMimeType}"
+    )
+
+    file_list = cognite_client.files.list(
+        metadata={docTypeMetaCol: pAndIdDocType},
+        data_set_external_ids=[docDataSetExtId],
+        mime_type=orgMimeType,
+        limit=docLimit,
+    )
```

Comment:
This seems risky/inefficient. You are calling a request in a loop that is independent of the loop variable, in other words, you are doing the same call to CDF for each root assets. Thus this will scale poorly.

Reply:
Agree: adding
asset_subtree_external_ids=[assetRootExtId],
to include the loop variable in the list command